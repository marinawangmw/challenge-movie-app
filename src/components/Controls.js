import React from 'react';
import { View, Text, StyleSheet, Image, TouchableOpacity } from 'react-native';
import { useTheme } from '@react-navigation/native';
import { spacing, typography } from '@/theme';
import { strings } from '@/localization';

export const Controls = ({ controlDatas, controlStyles }) => {
  const { colors } = useTheme();
  const styles = { ...defaultStyles, ...controlStyles };

  const renderControls = (control, idx) => (
    <View style={styles.controlWrapper} key={idx}>
      <TouchableOpacity onPress={control.handleControlPress}>
        <Image
          source={control.icon}
          style={[styles.icon, control.label === strings.controls.play && styles.playIcon]}
          resizeMode="cover"
          accessibilityIgnoresInvertColors
        />
      </TouchableOpacity>
      <Text style={[{ color: colors.text }, typography.label]}>{control.label}</Text>
    </View>
  );

  return (
    <View style={[styles.container, { color: colors.text }]}>
      {controlDatas.map((control, idx) => renderControls(control, idx))}
    </View>
  );
};

export const defaultStyles = StyleSheet.create({
  icon: {
    width: 28,
    height: 28,
    margin: spacing.xs,
  },
  playIcon: {
    height: 34,
    marginTop: spacing.xs,
    marginBottom: 2,
  },
});
