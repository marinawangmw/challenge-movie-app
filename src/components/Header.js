import React from 'react';
import { View, StyleSheet, Image } from 'react-native';
import { spacing } from '@/theme';

export const Header = () => {
  return (
    <View style={styles.header}>
      <Image
        source={require('@/assets/logo/logo.png')}
        style={styles.headerImg}
        resizeMode="contain"
        accessibilityIgnoresInvertColors
      />
    </View>
  );
};

const styles = StyleSheet.create({
  header: {
    marginTop: spacing.xl,
    alignItems: 'center',
    justifyContent: 'center',
  },
  headerImg: {
    height: 40,
    width: 76,
    marginTop: spacing.s,
  },
});
