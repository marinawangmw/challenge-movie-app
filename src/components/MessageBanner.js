import React from 'react';
import { View, Text } from 'react-native';
import { useTheme } from '@react-navigation/native';

export const MessageBanner = ({
  message = "There's no movie added to My List yet",
  customStyles,
}) => {
  const { colors } = useTheme();

  return (
    <View style={[customStyles.messageContainer]}>
      <Text style={{ color: colors.text }}>{message}</Text>
    </View>
  );
};
