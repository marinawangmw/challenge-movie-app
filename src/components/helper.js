import { Linking } from 'react-native';
import { Config } from 'react-native-config';
import Toast from 'react-native-toast-message';
import { getMovieTrailer } from '@/controllers/MovieController';

export const playTrailer = async (id) => {
  try {
    const trailerData = await getMovieTrailer(id);

    if (trailerData.results.length) {
      const url = Config.VIDEO_EXTERNAL_BASE_URL + trailerData.results[0].key;

      const supported = await Linking.canOpenURL(url);

      if (supported) {
        await Linking.openURL(url);
      } else {
        Toast.show({
          type: 'error',
          text1: "Something's wrong",
          text2: `Don't know how to open this URL: ${url}`,
        });
      }
    } else {
      Toast.show({
        type: 'error',
        text1: 'Oops',
        text2: "There's no trailer found for this movie",
      });
    }
  } catch (e) {
    console.log(e);
  }
};
