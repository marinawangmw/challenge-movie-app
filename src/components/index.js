export { Button } from '@/components/Button';
export { ErrorView } from '@/components/ErrorView';
export { TabBarIcon } from '@/components/TabBarIcon';
export { TextField } from '@/components/TextField';
export { Hero } from '@/components/Hero';
export { Header } from '@/components/Header';
export { MovieList } from '@/components/MovieList';
