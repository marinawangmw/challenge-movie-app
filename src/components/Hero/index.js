import React, { useEffect, useState, useCallback } from 'react';
import { View, ImageBackground, StyleSheet } from 'react-native';
import { useTheme } from '@react-navigation/native';
import { Config } from 'react-native-config';
import LinearGradient from 'react-native-linear-gradient';
import Labels from './Labels';
import SpecialBanner from './SpecialBanner';
import Controls from './Controls';
import { Header } from '@/components';
import { spacing } from '@/theme';
import { getTrending, getAllGenre } from '@/controllers/MovieController';

export const Hero = () => {
  const [posterURL, setPosterURL] = useState('');
  const [genreIds, setGenreIds] = useState([]);
  const [genres, setGenres] = useState([]);
  const [allGenres, setAllGenres] = useState(null);

  const { colors } = useTheme();

  const getData = useCallback(async () => {
    try {
      const responses = await Promise.all([getTrending(), getAllGenre()]);

      const posterMovieIndex = Math.floor(Math.random() * responses[0].results.length);
      const posterMovie = responses[0].results[posterMovieIndex];
      setPosterURL(posterMovie.poster_path);
      setGenreIds(posterMovie.genre_ids);

      const allGenresResult = responses[1];
      setAllGenres(allGenresResult);
    } catch (e) {
      console.log(e);
    }
  }, []);

  useEffect(() => {
    getData();
  }, [getData]);

  useEffect(() => {
    if (Boolean(allGenres) && genreIds.length) {
      setGenres(allGenres?.genres.filter((g) => genreIds.includes(g.id)));
    }
  }, [genreIds, allGenres]);

  return (
    <View style={[styles.container, { backgroundColor: colors.primary }]}>
      <ImageBackground
        source={{ uri: Config.IMAGE_API_BASE_URL + posterURL }}
        resizeMode="cover"
        style={styles.imgBg}
      >
        <LinearGradient
          locations={[0.0001, 1]}
          colors={['rgba(0, 0, 0, 0.0001)', '#000']}
          style={styles.gradient}
        >
          <Header />
          <View style={styles.bottomContainer}>
            <Labels labels={genres} />
            <SpecialBanner />
            <Controls />
          </View>
        </LinearGradient>
      </ImageBackground>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  gradient: {
    width: '100%',
    height: '100%',
  },
  imgBg: {
    width: '100%',
  },
  bottomContainer: {
    alignItems: 'center',
    alignSelf: 'center',
    position: 'absolute',
    bottom: 0,
    padding: spacing.s,
  },
});
