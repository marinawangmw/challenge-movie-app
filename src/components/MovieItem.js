import React from 'react';
import { StyleSheet, ImageBackground, Image, TouchableOpacity } from 'react-native';
import { Config } from 'react-native-config';
import { useTheme } from '@react-navigation/native';
import { spacing } from '@/theme';
import { movie } from '@/assets';
import { SpecialBanner } from '@/components';
import { NAVIGATION } from '@/constants';

export const MovieItem = ({ item, customStyles, navigation }) => {
  const { colors } = useTheme();
  const styles = { ...defaultStyles, ...customStyles };

  const handleNavigate = () => {
    navigation.navigate(NAVIGATION.movieDetail, {
      movieDetails: item.item,
    });
  };

  return (
    <>
      {item && item.item.poster_path ? (
        <TouchableOpacity onPress={handleNavigate} style={styles.poster}>
          <ImageBackground
            source={{ uri: Config.IMAGE_API_BASE_URL + item.item.poster_path }}
            resizeMode="cover"
            style={styles.img}
          >
            <SpecialBanner
              styles={{
                banner: { ...styles.banner, backgroundColor: colors.secondary },
                bannerLabel: [styles.bannerLabel, { color: colors.text }],
              }}
            />
          </ImageBackground>
        </TouchableOpacity>
      ) : (
        <TouchableOpacity
          style={[
            styles.img,
            styles.unknownContainer,
            { backgroundColor: colors.posterBackground },
          ]}
        >
          <Image
            source={movie}
            resizeMode="cover"
            style={styles.unknownPosterIcon}
            accessibilityIgnoresInvertColors
          />
        </TouchableOpacity>
      )}
    </>
  );
};

const defaultStyles = StyleSheet.create({
  title: {
    marginVertical: spacing.s,
  },
  img: {
    width: 90,
    height: 150,
    margin: 1,
  },
  unknownContainer: {
    justifyContent: 'center',
    alignItems: 'center',
  },
  unknownPosterIcon: {
    width: 50,
    height: 50,
  },
});
