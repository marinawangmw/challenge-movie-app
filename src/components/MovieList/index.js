import React from 'react';
import { StyleSheet, Text, FlatList, View, TouchableOpacity, Dimensions } from 'react-native';
import { useTheme } from '@react-navigation/native';
import { typography, spacing } from '@/theme';
import { MovieItem, MessageBanner } from '@/components';
import { NAVIGATION } from '@/constants';
import { en } from '@/localization/en';

const windowWidth = Dimensions.get('window').width;

export const MovieList = ({ navigation, item }) => {
  const { colors } = useTheme();

  const handlePress = () => {
    navigation.navigate(NAVIGATION.movieCollection, {
      collection: item,
      noObjectMessage: en.movieLists.noObjectMessage,
    });
  };

  const renderEmptyMessage = () => (
    <MessageBanner
      message={en.movieLists.noObjectMessage}
      customStyles={{
        messageContainer: { ...styles.messageContainer, backgroundColor: colors.card },
      }}
    />
  );

  return (
    <View style={styles.container}>
      <TouchableOpacity onPress={handlePress}>
        <Text style={[styles.title, { color: colors.text }, typography.title]}>{item.title}</Text>
      </TouchableOpacity>
      <FlatList
        data={item.movieList}
        renderItem={(movieItem) => <MovieItem item={movieItem} customStyles={movieItemStyles} />}
        horizontal
        ListEmptyComponent={renderEmptyMessage}
        contentContainerStyle={styles.flatList}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    margin: spacing.m,
  },
  title: {
    paddingVertical: spacing.s,
  },
  messageContainer: {
    height: 120,
    flexGrow: 1,
    marginVertical: spacing.xs,
    borderRadius: 4,
    justifyContent: 'center',
    alignItems: 'center',
  },
  flatList: {
    flexGrow: 1,
  },
});

const movieItemStyles = StyleSheet.create({
  poster: {
    width: 90,
    height: 150,
    margin: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  img: {
    width: 90,
    height: 150,
    margin: 1,
  },
  banner: {
    zIndex: 2,
    position: 'absolute',
    bottom: spacing.m,
    paddingHorizontal: spacing.xs,
    paddingVertical: 1,
    alignSelf: 'center',
    borderRadius: 1.25,
  },
  bannerLabel: {
    fontSize: 4.67,
    lineHeight: 5.47,
    fontWeight: 'bold',
  },
});
