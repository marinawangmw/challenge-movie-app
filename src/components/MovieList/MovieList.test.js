import React from 'react';
import { render } from '@testing-library/react-native';
import { MovieList } from '@/components';
import { withProviders } from '@/test-utils';

describe('Movie List renders correcly', () => {
  it('with empty movie list', () => {
    const item = { movieList: [], title: 'test' };
    const { toJSON } = render(withProviders(<MovieList item={item} />));
    expect(toJSON()).toMatchSnapshot();
  });

  it('with movie list with data', () => {
    const item = {
      movieList: [
        {
          adult: false,
          backdrop_path: '/ziTxBaF1AhbUwcv2IBiFJAf54qD.jpg',
          id: 65086,
          overview: 'The story follows a young lawyer, Arthur Kipps.',
          popularity: 16.334,
          poster_path: '/aJbeDPUwDe9MLpseTFrNTBaQPzf.jpg',
          release_date: '2012-02-03',
          title: 'The Woman in Black',
          vote_average: 6.11,
        },
        {
          adult: false,
          id: 102899,
          overview:
            'Armed with the astonishing ability to shrink in scale but increase in strength.',
          popularity: 69.704,
          poster_path: '/rS97hUJ1otKTTripGwQ0ujbuIri.jpg',
          release_date: '2015-07-14',
          title: 'Ant-Man',
          vote_average: 7.088,
        },
      ],
      title: 'test',
    };
    const { toJSON } = render(withProviders(<MovieList item={item} />));
    expect(toJSON()).toMatchSnapshot();
  });
});
