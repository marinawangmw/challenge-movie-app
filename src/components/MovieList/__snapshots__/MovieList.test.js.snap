// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie List renders correcly with empty movie list 1`] = `
<View
  style={
    Object {
      "margin": 24,
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "paddingVertical": 16,
          },
          Object {
            "color": "rgb(28, 28, 30)",
          },
          Object {
            "fontSize": 18,
            "fontWeight": "700",
          },
        ]
      }
    >
      test
    </Text>
  </View>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    contentContainerStyle={
      Object {
        "flexGrow": 1,
      }
    }
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgb(255, 255, 255)",
            "borderRadius": 4,
            "flexGrow": 1,
            "height": 120,
            "justifyContent": "center",
            "marginVertical": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "rgb(28, 28, 30)",
            }
          }
        >
          Nothing here yet
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Movie List renders correcly with movie list with data 1`] = `
<View
  style={
    Object {
      "margin": 24,
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "paddingVertical": 16,
          },
          Object {
            "color": "rgb(28, 28, 30)",
          },
          Object {
            "fontSize": 18,
            "fontWeight": "700",
          },
        ]
      }
    >
      test
    </Text>
  </View>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    contentContainerStyle={
      Object {
        "flexGrow": 1,
      }
    }
    data={
      Array [
        Object {
          "adult": false,
          "backdrop_path": "/ziTxBaF1AhbUwcv2IBiFJAf54qD.jpg",
          "id": 65086,
          "overview": "The story follows a young lawyer, Arthur Kipps.",
          "popularity": 16.334,
          "poster_path": "/aJbeDPUwDe9MLpseTFrNTBaQPzf.jpg",
          "release_date": "2012-02-03",
          "title": "The Woman in Black",
          "vote_average": 6.11,
        },
        Object {
          "adult": false,
          "id": 102899,
          "overview": "Armed with the astonishing ability to shrink in scale but increase in strength.",
          "popularity": 69.704,
          "poster_path": "/rS97hUJ1otKTTripGwQ0ujbuIri.jpg",
          "release_date": "2015-07-14",
          "title": "Ant-Man",
          "vote_average": 7.088,
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "height": 150,
              "justifyContent": "center",
              "margin": 1,
              "opacity": 1,
              "width": 90,
            }
          }
        >
          <View
            accessibilityIgnoresInvertColors={true}
            style={
              Object {
                "height": 150,
                "margin": 1,
                "width": 90,
              }
            }
          >
            <Image
              resizeMode="cover"
              source={
                Object {
                  "uri": "undefined/aJbeDPUwDe9MLpseTFrNTBaQPzf.jpg",
                }
              }
              style={
                Array [
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  Object {
                    "height": 150,
                    "width": 90,
                  },
                  undefined,
                ]
              }
            />
            <View
              style={
                Object {
                  "alignSelf": "center",
                  "backgroundColor": undefined,
                  "borderRadius": 1.25,
                  "bottom": 24,
                  "paddingHorizontal": 8,
                  "paddingVertical": 1,
                  "position": "absolute",
                  "zIndex": 2,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontSize": 4.67,
                      "fontWeight": "bold",
                      "lineHeight": 5.47,
                    },
                    Object {
                      "color": "rgb(28, 28, 30)",
                    },
                  ]
                }
              >
                Movy Original
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "height": 150,
              "justifyContent": "center",
              "margin": 1,
              "opacity": 1,
              "width": 90,
            }
          }
        >
          <View
            accessibilityIgnoresInvertColors={true}
            style={
              Object {
                "height": 150,
                "margin": 1,
                "width": 90,
              }
            }
          >
            <Image
              resizeMode="cover"
              source={
                Object {
                  "uri": "undefined/rS97hUJ1otKTTripGwQ0ujbuIri.jpg",
                }
              }
              style={
                Array [
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  Object {
                    "height": 150,
                    "width": 90,
                  },
                  undefined,
                ]
              }
            />
            <View
              style={
                Object {
                  "alignSelf": "center",
                  "backgroundColor": undefined,
                  "borderRadius": 1.25,
                  "bottom": 24,
                  "paddingHorizontal": 8,
                  "paddingVertical": 1,
                  "position": "absolute",
                  "zIndex": 2,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontSize": 4.67,
                      "fontWeight": "bold",
                      "lineHeight": 5.47,
                    },
                    Object {
                      "color": "rgb(28, 28, 30)",
                    },
                  ]
                }
              >
                Movy Original
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
